framework/src/postprocessors/GeneralSensorPostprocessor.C

```diff
@@ -0,0 +1,315 @@
+    std::sort(__time_values.begin(), __time_values.end());
+    std::sort(_for_exp.begin(), _for_exp.end());
+    integrand = elementwiseMultiply(__for_int, _for_exp);
+    // integrate the vector
+    //LinearInterpolation object with two vectors
+    LinearInterpolation integral(__time_values, integrand);
+    //integrate the vectors
+    integration_value = integral.integrate();
+  }
+
+  else 
+  {
+    integration_value = 0;
+  }
+
+  //-------------OUTPUT-------------
+  Real proportional_value = _input_signal_delayed + signalToNoise_value * 1;
+  sensor_value = drift_value + efficiency_value * (_proportional_weight * proportional_value + _integral_weight * integration_value);
+}
+
+PostprocessorValue
+GeneralSensorPostprocessor::getValue() const
+{  
+  return sensor_value;
+}
+
+void GeneralSensorPostprocessor::removeValuesBeforeTime(std::vector<Real>& _time_values, std::vector<Real>& _input_signal_values, std::vector<Real>& exp_values, Real t_desired) const 
+{
+  auto it = _time_values.begin();
+
+    // Find the iterator pointing to the element immediately before t_desired
+    auto lastBeforeTDesired = _time_values.begin();
+
+    while (it != _time_values.end())
+    {
+        if (*it < t_desired)
+        {
+            lastBeforeTDesired = it;
+            ++it;
+        }
+        else
+        {
+            // Break the loop when we find the first element greater than or equal to t_desired
+            break;
+        }
+    }
+
+    // Erase elements from the beginning after (and including) the iterator pointing to the element before t_desired
+    _time_values.erase(_time_values.begin(), lastBeforeTDesired+1);
+    _input_signal_values.erase(_input_signal_values.begin(), _input_signal_values.begin() + std::distance(_time_values.begin(), lastBeforeTDesired+1));
+    exp_values.erase(exp_values.begin(), lastBeforeTDesired+1);
+}
+
+Real GeneralSensorPostprocessor::linearInterpolation(Real x1, Real x2, Real y1, Real y2, Real x)
+{
+    Real y;
+    y = y1 + (x - x1) * (y2 - y1) / (x2 - x1);
+    return y;
+}
+
+Real GeneralSensorPostprocessor::linearInterpolationInVectors(std::vector<Real>& time, std::vector<Real>& y, Real desired_time)
+{
+    Real desired_y;
+    //----------begin-----------
+    // finding the index of t_1
+    Real t_1 = floor(desired_time/_dt)*_dt;
+    // Find the iterator for t_1
+    auto it_1 = find(time.begin(), time.end(), t_1);
+    // Find the index_1
+    size_t index_1 = distance(time.begin(), it_1);
+    Real fi_1 =  y[index_1];
+    // finding the index of t_2
+    Real t_2 = ceil(desired_time/_dt)*_dt;
+    // Find the iterator for t_2
+    auto it_2 = find(time.begin(), time.end(), t_2);
+    // Find the index_2
+    size_t index_2 = distance(time.begin(), it_2);
+    Real fi_2 =  y[index_2];
+    // Linear interpolation to calculate delayed input signal
+    desired_y = fi_1 + (desired_time - t_1) * (fi_2 - fi_1) / (t_2-t_1);
+    //------------end------------
+    return desired_y;

```

Comment:
Look at `LinearInterpolation.C` for a better way. This code assumes constant `_dt` over the whole simulation, which is not guaranteed.